### ISSUE TITLE
Large negative loss with InfoGAN on Yelp

### ISSUE THREAD

**Issue by thangld201** [dangne/tmd#3]
Hi @dangne, I tried training InfoGAN on Yelp for Roberta-base, but the losses are very negative. Is this normal ?
```
Epoch 22:   0%|                                                                       | 0/25 [00:00<?, ?it/s, loss=-52.8, v_num=8awv]
```

Also, did you use any validation metric for checkpoint selection, or you simply trained InfoGAN for 100 epochs and then selected the latest checkpoint ?

**Comment by dangne** [comment]
Hi @thangld201, thank you for your interest in this project.

1. Can you remind me which loss is this? This project was 2 years ago so I can't remember all the details anymore but I don't recall having any large negative loss?
2. Yes, I simply trained for 100 epochs and used the last one.

**Comment by thangld201** [comment]
Hi @dangne, it was the loss used to train the prior (p_loss)
```
        # Train the Prior Distribution
        if optimizer_idx == 2:
            # Compute Discriminator's predictions
            _, _, _, e_real_logits = self.dis(real_embs)  # (bs, code_dim)

            # Update scheduler
            self.pl_counter *= 0.999

            # Compute loss
            p_reg_loss = -(torch.softmax(self.prior, dim=-1) * torch.log_softmax(self.prior, dim=-1)).sum(dim=-1)
            p_ce_loss = -(torch.softmax(e_real_logits, dim=-1) * torch.log_softmax(self.prior, dim=-1)).sum(dim=-1).mean()
            p_loss = p_ce_loss - self.pl_counter * self.hparams.reg_prior_weight * p_reg_loss

            # Log results
            self.log(f"{self.state}/p_reg_loss", p_reg_loss)
            self.log(f"{self.state}/p_ce_loss", p_ce_loss)
            self.log(f"{self.state}/p_loss", p_loss)
```
You can view the curves [here](https://ibb.co/sWLLJCb)

**Comment by dangne** [comment]
I can't see any images in the link, but fortunately, I still have the wandb log. This is what the p_loss curve should look like for roberta+yelp:
<img width="343" alt="image" src="https://github.com/dangne/tmd/assets/36122564/fce2acd3-861b-4b34-94e9-59a8faf67e68">
Hope this helps.

**Comment by thangld201** [comment]
This makes sense. I am also seeing highly negative p_loss. I have some additional questions:
1. In the paper only the cross-entropy loss (p_ce_loss) was mentioned for prior learning, but I see in the codes that you also use entropy regularization (p_reg_loss). Is there a specific reason ?
2. By default, the downstream task LM is used to learn the manifold, but did you try learning the manifold with a raw LM (e.g. non-finetuned) and then projected upon a fine-tuned LM ? Just curious if you also did some preliminary exps.
3. I saw a number of other models implementations as well (vae, wgan) but seeing that you did not use them, probably they did not work or give very low results ?

**Comment by dangne** [comment]
1. At the start of the training, $Q(c|t)$ is still mostly random, and we don't want to use it to train our $r$ right away. That's why we introduce entropy regularization to maintain the randomness of $r$ at the initial steps. We then gradually decrease the regularization weight to allow $r$ to learn from $Q(c|t)$ in later steps.
2. Manifolds of the pretrained and fine-tuned LMs are likely to be different, so it does not make sense to approximate the manifold of one and then do projection on another.
3. Yes, I tried them and they did not work at all. The reason is that they cannot approximate disconnected manifolds; see Section 4.4 in our paper for the analysis. VAE, GAN, WGAN, and DCGAN are all connected-manifold models.

**Comment by dangne** [comment]
2. Although, a similar idea you can try to save compute resources is: when fine-tuning for downstream tasks, instead of tuning the entire LM, we only tune the classifier head. The pretrained LM is kept frozen and we approximate the manifold of this LM. I suppose this may be suboptimal in terms of classification accuracy but by doing this, we don't have to train a new InfoGAN for every downstream dataset.

**Comment by thangld201** [comment]
Thanks for your response. It explained things better!
